### PR TITLE
8329994: Zap alignment padding bits for ArrayOops in non-release builds

### PR DESCRIPTION
**Pull request by xmas92**
In a similar vein to [JDK-8325870](https://bugs.openjdk.org/browse/JDK-8325870). Zap the any alignment padding that may exist between the header and the payload.

Currently this padding only occurs when disabling `UseCompressedClassPointers`. But becomes more relevant with [JDK-8294992](https://bugs.openjdk.org/browse/JDK-8294992).

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8329994](https://bugs.openjdk.org/browse/JDK-8329994): Zap alignment padding bits for ArrayOops in non-release builds (**Enhancement** - P4)

### Reviewers
 * [Albert Mingkun Yang](https://openjdk.org/census#ayang) (@albertnetymk - **Reviewer**)
 * [Stefan Johansson](https://openjdk.org/census#sjohanss) (@kstefanj - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/18714/head:pull/18714` \
`$ git checkout pull/18714`

Update a local copy of the PR: \
`$ git checkout pull/18714` \
`$ git pull https://git.openjdk.org/jdk.git pull/18714/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 18714`

View PR using the GUI difftool: \
`$ git pr show -t 18714`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/18714.diff">https://git.openjdk.org/jdk/pull/18714.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/18714#issuecomment-2047324322)

### CODE REVIEW

File: src/hotspot/share/gc/shared/memAllocator.cpp [hunk 1]
```diff
@@ -406,13 +406,28 @@ oop ObjArrayAllocator::initialize(HeapWord* mem) const {
   assert(_length >= 0, "length should be non-negative");
   if (_do_zero) {
     mem_clear(mem);
+    mem_zap_alignment_padding(mem);
```
Comment (albertnetymk) [inline src/hotspot/share/gc/shared/memAllocator.cpp:409]: I feel the ordering here is a bit odd; these methods are for spaces corresponding to before-body, body and after-body.

File: src/hotspot/share/gc/shared/memAllocator.hpp [hunk 2]
```diff
@@ -98,6 +98,7 @@ class ObjArrayAllocator: public MemAllocator {
   const int  _length;
   const bool _do_zero;
 
+  void mem_zap_alignment_padding(HeapWord* mem) const PRODUCT_RETURN;
```
Comment (albertnetymk) [inline src/hotspot/share/gc/shared/memAllocator.hpp:101]: Both methods are about alignment-padding, one for array-element type, the other for obj itself. Maybe call it "mem_zap_start_padding" to make them more symmetric?

File: src/hotspot/share/gc/shared/memAllocator.cpp (hunk 1 above)
Comment (xmas92) [reply src/hotspot/share/gc/shared/memAllocator.cpp:409]: Yeah issue is that `mem_clear(mem);` will clear the padding. Maybe I should do the same for `ZObjArrayAllocator` as well. To keep them consistent. Even though the `ZObjArrayAllocator` does not clear that memory. 

Also this is debug code, maybe just doing it after the normal clearing is the better choice. So regardless how much memory the normal clearing does we always end up with the padding zapped.

File: src/hotspot/share/gc/shared/memAllocator.hpp (hunk 2 above)
Comment (xmas92) [reply src/hotspot/share/gc/shared/memAllocator.hpp:101]: I am not opposed to that name. Only confusion could be that someone thinks that it is before the header, but the current name is not any better. I'll take your suggestion.

**Review by albertnetymk — APPROVED** [review]

**Review by kstefanj — APPROVED** [review]

**Comment by xmas92** [comment]
Thanks for the reviews.
/integrate

**Pull request closed without merging**